### ISSUE TITLE
Move from VMR array to number density

### ISSUE THREAD

**Issue by AndrewJConley** [NCAR/Mechanism_Collection#3]
This requires changes to MechanismToCode, possibly to convergence tolerances, output units, input unit conversion, and who-knows-what-else.

**Comment by mattldawson** [comment]
Hi @AndrewJConley - a few quick questions:

- It looks like the species concentrations in the input file are volume mixing ratios. Will we assume that these will be in number concentration or should we plan on doing the conversion when the file is read?

- What will the output file units be?

- Are we using #/m3 or #/cc for number concentration? (My recommendation would be to stick with SI units and use #/m3)

**Comment by AndrewJConley** [comment]
I want to move to the #/m3 throughout the code and data (input and output).  We will need to write an extraction code as part of this effort to convert to this unit from the output in the files such as those on cheyenne (e.g., in /glade/scratch/shawnh/GEOS5_frcst_data_Beta/20200427/model_files/finn/

**Comment by mattldawson** [comment]
and the reactions in `mechanism.json` will also assume #/m3 units? (e.g., the Arrhenius pre-exponential factor)

**Comment by AndrewJConley** [comment]
Yes.  That will require a bit of work on my part.  And yet another reason not to have the preprocessor separate from the code.

**Comment by mattldawson** [comment]
ok, perfect.

**Comment by mattldawson** [comment]
Hi @AndrewJConley - I'm going to pass this back to you now. I've tried to adapt MusicBox, MICM and MechanismToCode to use number densities (#/m3) for chemical species.

The code compiles and runs (for me), but I haven't tried to verify the results because I think the last piece of this will be to adapt the chemistry cafe to provide rate constant parameters that assume concentrations in number density.

The changes I've made will need a lot of testing once this is done. In particular, the rate calculations in `src/kinetics/rate_constant_functions.F90` will need checked (particularly MCO3_NO2 and MPAN_M) to make sure the adjustments are correct.

There are `develop-3-number-density` branches in MICM_chemistry and MusicBox.

**Comment by AndrewJConley** [comment]
We have decided to move to moles/m^3.  So once the cafe and preprocessor produce code using those dimensions we will need to revisit those branches.

**Comment by AndrewJConley** [comment]
@mattldawson , Do I need to move to Moles/m^3?  Rather than #/m3?

**Comment by mattldawson** [comment]
I think we decided on moles/m3 because moles is a SI unit

**Comment by mattldawson** [comment]
outdated - we can create a new issue to review units in MusicBox-related repos if needed
